torch/csrc/autograd/saved_variable.cpp

```diff
@@ -153,27 +153,31 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
   return var;
 }
 
-void SavedVariable::register_hooks() {
+void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks) {
   if (!data_.defined()) {
     if (!was_default_constructed_) {
-    TORCH_CHECK(false,
-        "Calling register_hook on a saved tensor after it has been freed. "
+      TORCH_CHECK(false,
+        "Calling register_hooks on a saved tensor after it has been freed. "
         "Saved intermediate values "
         "of the graph are freed when you call .backward() or autograd.grad(). Specify "
         "retain_graph=True if you need to backward through the graph a second time or "
         "if you need to access saved variables after calling backward.");
     } else {
-    TORCH_CHECK(false, "Calling register_hook on a tensor with value None is forbidden");
+      TORCH_CHECK(false, "Calling register_hook on a tensor with value None is forbidden");
     }
   }
+  TORCH_CHECK(!hooks_,
+    "Calling register_hooks on a saved tensor whose hooks have already been set.");
+  hooks_ = std::move(hooks);
+  data_ = hooks_->call_unpack_hook(hooks_->call_pack_hook(data_));
```

Comment:
I think you will need to use a new PyObject field here.
Also you want to make sure to take a refcount to the returned pyobject to make sure it is not freed.

Reply:
Oh yeah sorry, that comes in the next PR

Reply:
I should have left a comment saying so

Reply:
Sorry, maybe I should add a comment explaining that this line is temporary and will disappear in #60975